File: 2019/utils/intcode.go
package utils

import (
	"fmt"
)

const (
	add         = 1
	multiply    = 2
	input       = 3
	output      = 4
	jumpIfTrue  = 5
	jumpIfFalse = 6
	lessThan    = 7
	equals      = 8
	adjust      = 9
	halt        = 99
)

var parameters = map[int]int{
	add:         3,
	multiply:    3,
	input:       1,
	output:      1,
	jumpIfTrue:  2,
	jumpIfFalse: 2,
	lessThan:    3,
	equals:      3,
	adjust:      1,
	halt:        0,
}

// RunIntCode runs the given instructions. The provided output is closed when the program halts.
func RunIntCode(in, out chan int, instructions []int) {
	defer close(out)

	memory := make(map[int]int)
	for i, instruction := range instructions {
		memory[i] = instruction
	}

	i := 0
	offset := 0
	for {
		opcode, ptrs := instruction(memory, i, offset)
		i += len(ptrs) + 1

		switch opcode {
		case add:
			memory[ptrs[2]] = memory[ptrs[0]] + memory[ptrs[1]]
		case multiply:
			memory[ptrs[2]] = memory[ptrs[0]] * memory[ptrs[1]]
		case input:
			memory[ptrs[0]] = <-in
		case output:
			out <- memory[ptrs[0]]
		case jumpIfTrue:
			if itob(memory[ptrs[0]]) {
				i = memory[ptrs[1]]
			}
		case jumpIfFalse:
			if !itob(memory[ptrs[0]]) {
				i = memory[ptrs[1]]
			}
		case lessThan:
			memory[ptrs[2]] = btoi(memory[ptrs[0]] < memory[ptrs[1]])
		case equals:
			memory[ptrs[2]] = btoi(memory[ptrs[0]] == memory[ptrs[1]])
		case adjust:
			offset += memory[ptrs[0]]
		case halt:
			return
		default:
			fmt.Printf("Unknown opcode %d (original %d), stopping!\n", opcode, memory[i])
			return
		}
	}
}

func instruction(memory map[int]int, pos, offset int) (int, []int) {
	opcode := memory[pos] % 100
	mask := memory[pos] / 100

	ptrs := make([]int, parameters[opcode])
	for i := 0; i < parameters[opcode]; i++ {
		switch mask % 10 {
		case 0:
			ptrs[i] = memory[pos+i+1]
		case 1:
			ptrs[i] = pos + i + 1
		case 2:
			ptrs[i] = memory[pos+i+1] + offset
		}
		mask = mask / 10
	}

	return opcode, ptrs
}

func btoi(b bool) int {
	if b {
		return 1
	}
	return 0
}

func itob(i int) bool {
	if i != 0 {
		return true
	}
	return false
}

File: 2019/kotlin/src/main/kotlin/nl/sanderp/aoc/aoc2019/IntCode.kt
package nl.sanderp.aoc.aoc2019

import kotlinx.coroutines.channels.ReceiveChannel
import kotlinx.coroutines.channels.SendChannel

internal enum class ParameterMode { Position, Immediate }

internal sealed class Instruction(val parameterCount: Int)
internal data class Add(val p1: Int, val p2: Int, val p3: Int) : Instruction(3)
internal data class Multiply(val p1: Int, val p2: Int, val p3: Int) : Instruction(3)
internal data class Input(val p1: Int) : Instruction(1)
internal data class Output(val p2: Int) : Instruction(1)
internal data class JumpIfTrue(val p1: Int, val p2: Int) : Instruction(2)
internal data class JumpIfFalse(val p1: Int, val p2: Int) : Instruction(2)
internal data class LessThan(val p1: Int, val p2: Int, val p3: Int): Instruction(3)
internal data class Equals(val p1: Int, val p2: Int, val p3: Int): Instruction(3)
internal object Halt : Instruction(0)

class IntCode(instructions: List<Int>) {
    private val memory = instructions.toIntArray()

    suspend fun run(input: ReceiveChannel<Int>, output: SendChannel<Int>) {
        var pointer = 0
        while (true) {
            val instruction = parseInstruction(pointer)
            pointer += instruction.parameterCount + 1
            when (instruction) {
                is Add -> memory[instruction.p3] = memory[instruction.p1] + memory[instruction.p2]
                is Multiply -> memory[instruction.p3] = memory[instruction.p1] * memory[instruction.p2]
                is Input -> memory[instruction.p1] = input.receive()
                is Output -> output.send(memory[instruction.p2])
                is JumpIfTrue -> {
                    if (memory[instruction.p1] != 0) {
                        pointer = memory[instruction.p2]
                    }
                }
                is JumpIfFalse -> {
                    if (memory[instruction.p1] == 0) {
                        pointer = memory[instruction.p2]
                    }
                }
                is LessThan -> memory[instruction.p3] = if (memory[instruction.p1] < memory[instruction.p2]) 1 else 0
                is Equals -> memory[instruction.p3] = if (memory[instruction.p1] == memory[instruction.p2]) 1 else 0
                is Halt -> break
            }
        }
        output.close()
    }

    private fun parseInstruction(pointer: Int): Instruction {
        val opcode = memory[pointer] % 100
        val parameters: Sequence<Int> = sequence {
            var mask = memory[pointer] / 100
            for (i in 1..3) {
                when (parseParameterMode(mask % 10)) {
                    ParameterMode.Immediate -> yield(pointer + i)
                    ParameterMode.Position -> yield(memory[pointer + i])
                }
                mask /= 10
            }
        }

        return when (opcode) {
            1 -> Add(parameters.elementAt(0), parameters.elementAt(1), parameters.elementAt(2))
            2 -> Multiply(parameters.elementAt(0), parameters.elementAt(1), parameters.elementAt(2))
            3 -> Input(parameters.first())
            4 -> Output(parameters.first())
            5 -> JumpIfTrue(parameters.elementAt(0), parameters.elementAt(1))
            6 -> JumpIfFalse(parameters.elementAt(0), parameters.elementAt(1))
            7 -> LessThan(parameters.elementAt(0), parameters.elementAt(1), parameters.elementAt(2))
            8 -> Equals(parameters.elementAt(0), parameters.elementAt(1), parameters.elementAt(2))
            99 -> Halt
            else -> throw IllegalArgumentException("Unknown opcode '$opcode'")
        }
    }

    private fun parseParameterMode(digit: Int) = when (digit) {
        0 -> ParameterMode.Position
        1 -> ParameterMode.Immediate
        else -> throw IllegalArgumentException("Unknown parameter mode '$digit'")
    }
}
